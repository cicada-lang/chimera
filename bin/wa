#!/usr/bin/env -S deno run --allow-all

import { createCommandRunner } from "../src/console/index.ts"

createCommandRunner().run()
