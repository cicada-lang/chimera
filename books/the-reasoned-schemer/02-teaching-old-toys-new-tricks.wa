Car [p, a]
---------- {
  [a, ...d] = p
}

query q {
  Car [["a", "c", "o", "r", "n"], q]
}

query q {
  Car [["a", "c", "o", "r", "n"], "a"]
}

query r {
  Car [[r, ...y], x]
  "pear" = x
}

query r {
  Car [["grape", "raisin", "pear"], x]
  Car [[["a"], ["b"], ["c"]], y]
  [x, ...y] = r
}

Cdr [p, d]
---------- {
  [a, ...d] = p
}

query r {
  Cdr [["a", "c", "o", "r", "n"], v]
  Cdr [v, w]
  Car [w, r]
}

query r {
  Cdr [["grape", "raisin", "pear"], x]
  Car [[["a"], ["b"], ["c"]], y]
  [x, ...y] = r
}

query q {
  Cdr [["a", "c", "o", "r", "n"], ["c", "o", "r", "n"]]
}

query x {
  Cdr [["c", "o", "r", "n"], [x, "r", "n"]]
}

query l {
  Cdr [l, ["c", "o", "r", "n"]]
  Car [l, x]
  x = "a"
}

// Cons [a, d, p]
// --------------- {
//   [a, ...d] = p
// }

Cons [a, d, p]
--------------- {
  Car [p, a]
  Cdr [p, d]
}

query l {
  Cons [["a", "b", "c"], ["d", "e"], l]
}

query x {
  Cons [x, ["a", "b", "c"], ["d", "a", "b", "c"]]
}

query r {
  ["e", "a", "d", x] = r
  Cons [y, ["a", z, "c"], r]
}

query x {
  Cons [x, ["a", x, "c"], ["d", "a", x, "c"]]
}

query l {
  ["d", "a", x, "c"] = l
  Cons [x, ["a", x, "c"], l]
}

query l {
  Cons [x, ["a", x, "c"], l]
  ["d", "a", x, "c"] = l
}

query l {
  Cons [w, ["n", "u", "s"], t]
  Cdr [l, t]
  Car [l, x]
  "b" = x
  Cdr [l, d]
  Car [d, y]
  "o" = y
}

Null x
---------- {
  [] = x
}

query q {
  Null ["grape", "raisin", "pear"]
}

query q {
  Null []
}

query x {
  Null x
}

query r {
  [x, y, "salad"] = r
}

Pair p
------ {
  Cons [a, d, p]
}

query q {
  Pair [q, ...q]
}

query q {
  Pair []
}

query q {
  Pair "pair"
}

query x {
  Pair x
}

query r {
  Pair [r]
}

// Singleton l
// ----------- {
//   Pair l
//   Cdr [l, d]
//   Null d
// }

Singleton l
----------- {
  Cdr [l, d]
  Null d
}

query r {
  Singleton ["tofu"]
}

query r {
  Singleton [["tofu"]]
}

query r {
  Singleton "tofu"
}

query r {
  Singleton ["e", "tofu"]
}

query r {
  Singleton []
}

query r {
  Singleton ["e", ..."tofu"]
}
