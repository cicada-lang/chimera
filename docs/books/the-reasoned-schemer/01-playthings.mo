export clause Succeed()

export clause Fail()
-------- {
  1 = 2
}

print find _ {
  Fail()
}

print find _ {
  "pea" = "pod"
}

print find q {
  q = "pea"
}

print find q {
  "pea" = q
}

print find _ {
  Succeed()
}

print find _ {
  "pea" = "pea"
}

print find q {
  q = q
}

print find q {
  q = _
}

print find q {
  q = [_]
}

print find _ {
  [[["pea"]], "pod"] = [[["pea"]], "pod"]
}

print find q {
  [[[q]], "pod"] = [[[_]], "pod"]
}

print find q {
  [[[q]], x] = [[[x]], "pod"]
}

print find q {
  [x, x] = q
}

print find q {
  [_x, _y] = q
}

print find s {
  [_t, _u] = s
}

print find q {
  [x, _y, x] = q
}

print find _ {
  ["pea"] = "pea"
}

print find _ {
  [x] = x
}

print find _ {
  Succeed()
  Succeed()
}

print find q {
  Succeed()
  "corn" = q
}

print find q {
  Fail()
  "corn" = q
}

print find q {
  "corn" = q
  Fail()
}

print find q {
  "corn" = q
  "meal" = q
}

print find q {
  "corn" = q
  "corn" = q
}

clause Teacup(t)
--------------- tea {
  t = "tea"
}

clause Teacup(t)
--------------- cup {
  t = "cup"
}

print find q {
  Teacup(q)
}

print find [x, y] {
  Teacup(x)
  Teacup(y)
}

print find [x, _y] {
  Teacup(x)
  Teacup(x)
}

print find [x, _y] {
  Teacup(x)
}
