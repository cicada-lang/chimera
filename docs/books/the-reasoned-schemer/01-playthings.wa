Succeed []

Fail []
-------- {
  1 = 2
}

find _ {
  Fail []
}

find _ {
  "pea" = "pod"
}

find q {
  q = "pea"
}

find q {
  "pea" = q
}

find _ {
  Succeed []
}

find _ {
  "pea" = "pea"
}

find q {
  q = q
}

find q {
  q = _
}

find q {
  q = [_]
}

find _ {
  [[["pea"]], "pod"] = [[["pea"]], "pod"]
}

find q {
  [[[q]], "pod"] = [[[_]], "pod"]
}

find q {
  [[[q]], x] = [[[x]], "pod"]
}

find q {
  [x, x] = q
}

find q {
  [_x, _y] = q
}

find s {
  [_t, _u] = s
}

find q {
  [x, _y, x] = q
}

find _ {
  ["pea"] = "pea"
}

find _ {
  [x] = x
}

find _ {
  Succeed []
  Succeed []
}

find q {
  Succeed []
  "corn" = q
}

find q {
  Fail []
  "corn" = q
}

find q {
  "corn" = q
  Fail []
}

find q {
  "corn" = q
  "meal" = q
}

find q {
  "corn" = q
  "corn" = q
}

Teacup [t]
--------------- tea {
  t = "tea"
}

Teacup [t]
--------------- cup {
  t = "cup"
}

find q {
  Teacup [q]
}

find [x, y] {
  Teacup [x]
  Teacup [y]
}

find [x, _y] {
  Teacup [x]
  Teacup [x]
}

find [x, _y] {
  Teacup [x]
}
