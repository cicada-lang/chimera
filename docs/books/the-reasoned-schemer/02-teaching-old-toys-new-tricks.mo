export clause Car(p, a)
---------- {
  Equal([a | _d], p)
}

print find q {
  Car(["a", "c", "o", "r", "n"], q)
}

print find _ {
  Car(["a", "c", "o", "r", "n"], "a")
}

print find r {
  Car([r | _y], x)
  Equal("pear", x)
}

print find r {
  Car(["grape", "raisin", "pear"], x)
  Car([["a"], ["b"], ["c"]], y)
  Equal([x | y], r)
}

export clause Cdr(p, d)
---------- {
  Equal([_a | d], p)
}

print find r {
  Cdr(["a", "c", "o", "r", "n"], v)
  Cdr(v, w)
  Car(w, r)
}

print find r {
  Cdr(["grape", "raisin", "pear"], x)
  Car([["a"], ["b"], ["c"]], y)
  Equal([x | y], r)
}

print find _ {
  Cdr(["a", "c", "o", "r", "n"], ["c", "o", "r", "n"])
}

print find x {
  Cdr(["c", "o", "r", "n"], [x, "r", "n"])
}

print find l {
  Cdr(l, ["c", "o", "r", "n"])
  Car(l, x)
  Equal(x, "a")
}

// clause Cons(a, d, p)
// --------------- {
//   Equal([a | d], p)
// }

export clause Cons(a, d, p)
--------------- {
  Car(p, a)
  Cdr(p, d)
}

print find l {
  Cons(["a", "b", "c"], ["d", "e"], l)
}

print find x {
  Cons(x, ["a", "b", "c"], ["d", "a", "b", "c"])
}

print find r {
  Equal(["e", "a", "d", _x], r)
  Cons(_y, ["a", _z, "c"], r)
}

print find x {
  Cons(x, ["a", x, "c"], ["d", "a", x, "c"])
}

print find l {
  Equal(["d", "a", x, "c"], l)
  Cons(x, ["a", x, "c"], l)
}

print find l {
  Cons(x, ["a", x, "c"], l)
  Equal(["d", "a", x, "c"], l)
}

print find l {
  Cons(_w, ["n", "u", "s"], t)
  Cdr(l, t)
  Car(l, x)
  Equal("b", x)
  Cdr(l, d)
  Car(d, y)
  Equal("o", y)
}

export clause Null(x)
---------- {
  Equal([], x)
}

print find _ {
  Null(["grape", "raisin", "pear"])
}

print find _ {
  Null([])
}

print find x {
  Null(x)
}

print find r {
  Equal([_x, _y, "salad"], r)
}

clause Pair(p)
------ {
  Cons(_a, _d, p)
}

print find q {
  Pair([q | q])
}

print find _ {
  Pair([])
}

print find _ {
  Pair("pair")
}

print find x {
  Pair(x)
}

print find r {
  Pair([r])
}

// clause Singleton(l)
// ----------- {
//   Pair(l)
//   Cdr(l, d)
//   Null(d)
// }

clause Singleton(l)
----------- {
  Cdr(l, d)
  Null(d)
}

print find _ {
  Singleton(["tofu"])
}

print find _ {
  Singleton([["tofu"]])
}

print find _ {
  Singleton("tofu")
}

print find _ {
  Singleton(["e", "tofu"])
}

print find _ {
  Singleton([])
}

print find _ {
  Singleton(["e" | "tofu"])
}
