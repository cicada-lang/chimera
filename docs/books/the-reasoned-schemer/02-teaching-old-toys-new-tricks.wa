Car [p, a]
---------- {
  [a | _d] = p
}

find q {
  Car [["a", "c", "o", "r", "n"], q]
}

find q {
  Car [["a", "c", "o", "r", "n"], "a"]
}

find r {
  Car [[r | y], x]
  "pear" = x
}

find r {
  Car [["grape", "raisin", "pear"], x]
  Car [[["a"], ["b"], ["c"]], y]
  [x | y] = r
}

Cdr [p, d]
---------- {
  [_a | d] = p
}

find r {
  Cdr [["a", "c", "o", "r", "n"], v]
  Cdr [v, w]
  Car [w, r]
}

find r {
  Cdr [["grape", "raisin", "pear"], x]
  Car [[["a"], ["b"], ["c"]], y]
  [x | y] = r
}

find q {
  Cdr [["a", "c", "o", "r", "n"], ["c", "o", "r", "n"]]
}

find x {
  Cdr [["c", "o", "r", "n"], [x, "r", "n"]]
}

find l {
  Cdr [l, ["c", "o", "r", "n"]]
  Car [l, x]
  x = "a"
}

// Cons [a, d, p]
// --------------- {
//   [a | d] = p
// }

Cons [a, d, p]
--------------- {
  Car [p, a]
  Cdr [p, d]
}

find l {
  Cons [["a", "b", "c"], ["d", "e"], l]
}

find x {
  Cons [x, ["a", "b", "c"], ["d", "a", "b", "c"]]
}

find r {
  ["e", "a", "d", x] = r
  Cons [y, ["a", z, "c"], r]
}

find x {
  Cons [x, ["a", x, "c"], ["d", "a", x, "c"]]
}

find l {
  ["d", "a", x, "c"] = l
  Cons [x, ["a", x, "c"], l]
}

find l {
  Cons [x, ["a", x, "c"], l]
  ["d", "a", x, "c"] = l
}

find l {
  Cons [w, ["n", "u", "s"], t]
  Cdr [l, t]
  Car [l, x]
  "b" = x
  Cdr [l, d]
  Car [d, y]
  "o" = y
}

Null x
---------- {
  [] = x
}

find q {
  Null ["grape", "raisin", "pear"]
}

find q {
  Null []
}

find x {
  Null x
}

find r {
  [x, y, "salad"] = r
}

Pair p
------ {
  Cons [_a, _d, p]
}

find q {
  Pair [q | q]
}

find q {
  Pair []
}

find q {
  Pair "pair"
}

find x {
  Pair x
}

find r {
  Pair [r]
}

// Singleton l
// ----------- {
//   Pair l
//   Cdr [l, d]
//   Null d
// }

Singleton l
----------- {
  Cdr [l, d]
  Null d
}

find r {
  Singleton ["tofu"]
}

find r {
  Singleton [["tofu"]]
}

find r {
  Singleton "tofu"
}

find r {
  Singleton ["e", "tofu"]
}

find r {
  Singleton []
}

find r {
  Singleton ["e" | "tofu"]
}
