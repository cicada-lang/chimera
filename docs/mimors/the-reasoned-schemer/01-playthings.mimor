<theme color="orange" />

<question>
  What is a goal?

  <answer>
    It is something that either succeeds, fails, or has no value.
  </answer>
</question>

<question>
  What is a fresh variable?

  <answer>
    Every variable is initially fresh.

    A variable is no longer fresh
    if it becomes associated with a non-variable value
    or if it becomes associated with a variable that,
    itself, is no longer fresh.
  </answer>
</question>

<question>
  When are two variables different?

  <answer>
    Two variables are different if they have not been fused.

    Every variable introduced by fresh (or run*)
    is initially different from every other variable.
  </answer>
</question>

<question>
  What does it mean for x to occur?

  <answer>
    A variable x occurs in a variable y
    when x (or any variable fused with x)
    appears in the value associated with y.
  </answer>
</question>

<question>
  When do we say a variable occurs in a list?

  <answer>
    A variable x occurs in a list l
    when x (or any variable fused with x) is an element of l,
    or when x occurs in an element of l.
  </answer>
</question>

<question>
  What is a relation?

  <answer>
    A relation is a kind of function that,
    when given arguments, produces a goal.

    Thanks, Robert A. Kowalski (1941 --).
  </answer>
</question>

<question>
  The Second Law of unify

  <answer>
    If x is fresh, then

    <code>
      unify v = x
    </code>

    succeeds and associates v with x,
    unless x occurs in v.
  </answer>
</question>

<question>
  The First Law of unify

  <answer>
    <code>
      unify v = w
    </code>

    can be replaced by

    <code>
      unify w = v
    </code>
  </answer>
</question>

<question>
  What does the "e" in conde stand for?

  <answer>
    It stands for every,
    since every successful conde line
    contributes one or more values (to the solution).
  </answer>
</question>
