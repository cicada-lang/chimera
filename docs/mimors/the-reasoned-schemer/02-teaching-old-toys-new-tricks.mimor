<theme color="blue" />

<question>
  How do we translate a function into a relation?

  <answer>
    The Translation (Initial)

    To translate a function into a relation,
    first replace define with defrel.

    Then unnest each expression in each cond line,
    and replace each cond with conde.

    To unnest a #t, replace it with #s.

    To unnest a #f, replace it with #u.
  </answer>
</question>

<question>
  The Law of #u (fail)

  <answer>
    Any conde line that has #u as a top-level goal cannot contribute values.
  </answer>
</question>
