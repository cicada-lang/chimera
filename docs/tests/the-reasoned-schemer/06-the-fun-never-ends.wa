import { Succeed, Fail } from "./01-playthings.wa"
import { Null, Cons, Car, Cdr } from "./02-teaching-old-toys-new-tricks.wa"

Always []
--------- {
  Succeed []
}

Always []
--------- {
  Always []
}

query q limit 3 {
  Always []
}

query q limit 3 {
  unify q = "onion"
  Always []
}

query q limit 3 {
  Always []
  unify q = "onion"
}

query q {
  Always []
  Fail []
}

query q {
  unify q = "garlic"
  Always []
  unify q = "onion"
}

Never []
--------- {
  Never []
}

query q {
  Fail []
  Never []
}

query q {
  Never []
  Fail []
}

query q {
  Never []
  Fail []
  Never []
}
