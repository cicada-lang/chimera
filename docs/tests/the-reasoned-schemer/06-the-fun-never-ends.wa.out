[]
[]
[ "pea" ]
[ "pea" ]
[ "_.0" ]
[ "_.0" ]
[ "_.0" ]
[ "_.0" ]
[ ["_.0"] ]
[ "_.0" ]
[ "_.0" ]
[ "pod" ]
[ ["_.0", "_.0"] ]
[ ["_.0", "_.1"] ]
[ ["_.0", "_.1"] ]
[ ["_.0", "_.1", "_.0"] ]
[]
[]
[ "_.0" ]
[ "corn" ]
[]
[]
[]
[ "corn" ]
[ "tea", 
  "cup" ]
[ ["tea", "tea"], 
  ["tea", "cup"], 
  ["cup", "tea"], 
  ["cup", "cup"] ]
[ ["tea", "_.0"], 
  ["cup", "_.0"] ]
[ ["tea", "_.0"], 
  ["cup", "_.0"] ]
[ "a" ]
[ "_.0" ]
[ "pear" ]
[ ["grape", "a"] ]
[ "o" ]
[ [["raisin", "pear"], "a"] ]
[ "_.0" ]
[ "o" ]
[ ["a", "c", "o", "r", "n"] ]
[ [["a", "b", "c"], "d", "e"] ]
[ "d" ]
[ ["e", "a", "d", "c"] ]
[ "d" ]
[ ["d", "a", "d", "c"] ]
[ ["d", "a", "d", "c"] ]
[ ["b", "o", "n", "u", "s"] ]
[]
[ "_.0" ]
[ [] ]
[ ["_.0", "_.1", "salad"] ]
[ "_.0" ]
[]
[]
[ ["_.0", { "...": "_.1" }] ]
[ "_.0" ]
[ "_.0" ]
[ "_.0" ]
[]
[]
[]
[]
[ "_.0", 
  "_.0", 
  "_.0" ]
[ "onion", 
  "onion", 
  "onion" ]
[ "onion", 
  "onion", 
  "onion" ]
[]
[]
[]
[]
[]
